# PRACTICA DE PANDAS <a class="tocSkip">
## Curso Programacion Analitica <a class="tocSkip">
### Maestria TIC Linea Ciencia de Datos <a class="tocSkip">
### Por [Jose R. Zapata](https://joserzapata.github.io/)  <a class="tocSkip">
joser.zapata@upb.edu.co

**Ejercicio SF Salaries**
Utilizaremos una variacion del SF Salaries Dataset de kaggle. Simplemente siga y complete las tareas que se detallan a continuación en negrita. Las tareas serán cada vez más difíciles a medida que avance.

La idea principal es aplicar pandas para organizar informacion, manipularla, buscar informacion en el dataset


**Importe pandas como pd.**

**lea Salaries.csv como un dataframe llamado sal.**

**Observe los primeros 5 registros del DataFrame.**

In [0]:
#Copie el Codigo Aca


Id       EmployeeName                                        JobTitle  \
0   1     NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   
1   2       GARY JIMENEZ                 CAPTAIN III (POLICE DEPARTMENT)   
2   3     ALBERT PARDINI                 CAPTAIN III (POLICE DEPARTMENT)   
3   4  CHRISTOPHER CHONG            WIRE ROPE CABLE MAINTENANCE MECHANIC   
4   5    PATRICK GARDNER    DEPUTY CHIEF OF DEPARTMENT,(FIRE DEPARTMENT)   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18         0.00  400184.25       NaN  567595.43         567595.43   
1  155966.02    245131.88  137811.38       NaN  538909.28         538909.28   
2  212739.13    106088.18   16452.60       NaN  335279.91         335279.91   
3   77916.00     56120.71  198306.90       NaN  332343.61         332343.61   
4  134401.60      9737.00  182234.59       NaN  326373.19         326373.19   

   Year  Notes         Agency  Status  
0  2011    NaN  San Francisco     NaN  
1  2011    NaN  San Francisco     NaN  
2  2011    NaN  San Francisco     NaN  
3  2011    NaN  San Francisco     NaN  
4  2011    NaN  San Francisco     NaN

**Observe los ultimos 5 registros del DataFrame.**

In [0]:
#Copie el Codigo Aca


Id   EmployeeName                    JobTitle  BasePay  \
148649  148650  Roy I Tillery                   Custodian      0.0   
148650  148651   Not provided                Not provided      NaN   
148651  148652   Not provided                Not provided      NaN   
148652  148653   Not provided                Not provided      NaN   
148653  148654      Joe Lopez  Counselor, Log Cabin Ranch      0.0   

        OvertimePay  OtherPay  Benefits  TotalPay  TotalPayBenefits  Year  \
148649          0.0      0.00       0.0      0.00              0.00  2014   
148650          NaN       NaN       NaN      0.00              0.00  2014   
148651          NaN       NaN       NaN      0.00              0.00  2014   
148652          NaN       NaN       NaN      0.00              0.00  2014   
148653          0.0   -618.13       0.0   -618.13           -618.13  2014   

               Agency  
148649  San Francisco  
148650  San Francisco  
148651  San Francisco  
148652  San Francisco  
148653  San Francisco

**¿Cuanto es el numero de filas y columnas del dataframe?**

In [0]:
#Copie el Codigo Aca


(148654, 13)

**Use el metodo `.info()` para encontrar cuantas Columnas de atributos hay y que tipo de datos tiene cada una.**

In [0]:
#Copie el Codigo Aca


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 13 columns):
Id                  148654 non-null int64
EmployeeName        148654 non-null object
JobTitle            148654 non-null object
BasePay             148045 non-null float64
OvertimePay         148650 non-null float64
OtherPay            148650 non-null float64
Benefits            112491 non-null float64
TotalPay            148654 non-null float64
TotalPayBenefits    148654 non-null float64
Year                148654 non-null int64
Notes               0 non-null float64
Agency              148654 non-null object
Status              0 non-null float64
dtypes: float64(8), int64(2), object(3)
memory usage: 14.7+ MB


- Segun la informacion del dataframe su tamaño es de *memory usage: 14.7+ MB*

- las columnas **Notes** y **Status** solamente tienen valores *NaN*, verifique que dice **0 non-null** osea 0 valores son diferentes a *NaN*
entonces todos los valores son *NaN* en esas columnas

**Elimine las columnas *Notes* y *Status* Recuerde usar el metodo de forma permanente (inplace)**

**Observe los 25 registros que estan desde el indice 3000**

In [0]:
#Copie el Codigo Aca


Id          EmployeeName                            JobTitle  \
3000  3001         WILLIAM BOYLE                         FIREFIGHTER   
3001  3002              ALAN LIN   ELECTRONIC MAINTENANCE TECHNICIAN   
3002  3003        WILLIAM CONLEY    SERGEANT III (POLICE DEPARTMENT)   
3003  3004       ROBERT MCDONALD         LIEUTENANT, FIRE DEPARTMENT   
3004  3005      MICHAEL COSTELLO                  POLICE OFFICER III   
3005  3006               EDNA HO                           MANAGER V   
3006  3007         RODRIGO CONEL                    REGISTERED NURSE   
3007  3008        VICTOR HURTADO                         FIREFIGHTER   
3008  3009         DONALD LANDRY           TRANSIT POWER LINE WORKER   
3009  3010             FRANK LEE  INSPECTOR III, (POLICE DEPARTMENT)   
3010  3011           GARY CASTEL                  POLICE OFFICER III   
3011  3012           JAKE PETERS      TRANSMISSION LINE SUPERVISOR I   
3012  3013     NATALIE LEVCHENKO              FIRE FIGHTER PARAMEDIC   
3013  3014        WILLIAM LINNEY                         FIREFIGHTER   
3014  3015  JANETTE NEVES RIVERA                         FIREFIGHTER   
3015  3016           SOPHAL CHEA                  POLICE OFFICER III   
3016  3017          HERMAN JONES  INSPECTOR III, (POLICE DEPARTMENT)   
3017  3018       RAYMOND MATTIAS             SEWER REPAIR SUPERVISOR   
3018  3019       JOSEPH KAVANAGH                   POLICE OFFICER II   
3019  3020        SANGEETA SINHA           ATTORNEY (CIVIL/CRIMINAL)   
3020  3021   ROBERTO EVANGELISTA           ATTORNEY (CIVIL/CRIMINAL)   
3021  3022           RICHARD KIM                  POLICE OFFICER III   
3022  3023       ROBERT COSTELLO        UTILITY PLUMBER SUPERVISOR I   
3023  3024           STANLEY LEE                  POLICE OFFICER III   
3024  3025           STANLEY LEE                         FIREFIGHTER   

        BasePay  OvertimePay  OtherPay  Benefits   TotalPay  TotalPayBenefits  \
3000  105934.65     18530.98  16389.60       NaN  140855.23         140855.23   
3001   98965.80     41504.40    380.00       NaN  140850.20         140850.20   
3002  130457.78      3131.42   7256.39       NaN  140845.59         140845.59   
3003  123104.99      1833.56  15897.68       NaN  140836.23         140836.23   
3004  112401.08      7957.74  20466.93       NaN  140825.75         140825.75   
3005  140821.37         0.00      0.00       NaN  140821.37         140821.37   
3006  126120.00      3981.78  10714.21       NaN  140815.99         140815.99   
3007  105934.67     13112.76  21763.50       NaN  140810.93         140810.93   
3008   92861.50     39789.32   8135.86       NaN  140786.68         140786.68   
3009  130481.56       725.40   9579.58       NaN  140786.54         140786.54   
3010  117303.67     11019.70  12457.26       NaN  140780.63         140780.63   
3011   97198.24     42162.36   1406.88       NaN  140767.48         140767.48   
3012  121431.62      7821.61  11508.63       NaN  140761.86         140761.86   
3013  105934.67     12365.94  22460.59       NaN  140761.20         140761.20   
3014  100280.70     30670.65   9777.87       NaN  140729.22         140729.22   
3015  111038.42     10436.21  19223.74       NaN  140698.37         140698.37   
3016  130457.92      6185.03   4054.37       NaN  140697.32         140697.32   
3017   93563.60     28298.93  18830.13       NaN  140692.66         140692.66   
3018  109112.08     19102.62  12477.29       NaN  140691.99         140691.99   
3019  140691.97         0.00      0.00       NaN  140691.97         140691.97   
3020  140691.95         0.00      0.00       NaN  140691.95         140691.95   
3021  112421.44     18160.54  10105.94       NaN  140687.92         140687.92   
3022  102050.03     32089.98   6544.93       NaN  140684.94         140684.94   
3023  112134.66      3373.46  25169.78       NaN  140677.90         140677.90   
3024  105934.67     14158.45  20577.18       NaN  140670.30         140670.30   

      Year         Agency  


**¿Cual es la nueva informacion del dataframe?**

In [0]:
#Copie el Codigo Aca


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
Id                  148654 non-null int64
EmployeeName        148654 non-null object
JobTitle            148654 non-null object
BasePay             148045 non-null float64
OvertimePay         148650 non-null float64
OtherPay            148650 non-null float64
Benefits            112491 non-null float64
TotalPay            148654 non-null float64
TotalPayBenefits    148654 non-null float64
Year                148654 non-null int64
Agency              148654 non-null object
dtypes: float64(6), int64(2), object(3)
memory usage: 12.5+ MB


- Se puede verificar que al borrar las columnas que no tenian solo *NaN* se disminuye el tamaño del dataframe a: *memory usage: 12.5+ MB*

- La columna *Benefits* tiene valores *NaN, ¿Cuantos NaN tiene la columna *Benefits*? (TIP: Seleccione la Columna y vefique cuanots Na tiene `.isna()` luego sumelos `.sum()` )

In [0]:
#Copie el Codigo Aca


36163

Al parecer los valores faltantes en la columna Benefits es por que no se le pagan beneficios entonces:

**Reemplace los valores *NaN* de la Columna *Benefits* por el valor de cero (Recuerde usar el metodo de forma inplace) y mostrar los primeros 10 elementos de la lista**

**¿Cual es la nueva informacion del dataframe?**

In [0]:
#Copie el Codigo Aca


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 148654 entries, 0 to 148653
Data columns (total 11 columns):
Id                  148654 non-null int64
EmployeeName        148654 non-null object
JobTitle            148654 non-null object
BasePay             148045 non-null float64
OvertimePay         148650 non-null float64
OtherPay            148650 non-null float64
Benefits            148654 non-null float64
TotalPay            148654 non-null float64
TotalPayBenefits    148654 non-null float64
Year                148654 non-null int64
Agency              148654 non-null object
dtypes: float64(6), int64(2), object(3)
memory usage: 12.5+ MB


**Cual es el promedio de BasePay ?**

In [0]:
#Copie el Codigo Aca


66325.44884050643

**Cual es la cantidad mas alta de OvertimePay en el dataset ?**

In [0]:
#Copie el Codigo Aca


245131.88

**¿Cuál es el título de trabajo de JOSEPH DRISCOLL? Nota: Utilice mayúsculas, de lo contrario, puede obtener una respuesta que no coincida.**

In [0]:
#Copie el Codigo Aca


24    CAPTAIN, FIRE SUPPRESSION
Name: JobTitle, dtype: object

**¿Cuanto gana JOSEPH DRISCOLL (incluyendo beneficios)?**

In [0]:
#Copie el Codigo Aca


24    270324.91
Name: TotalPayBenefits, dtype: float64

**¿Cual es el nombre de la persona con mejor pago (incluyendo beneficios)?**

In [0]:
#Copie el Codigo Aca


Id    EmployeeName                                        JobTitle  \
0   1  NATHANIEL FORD  GENERAL MANAGER-METROPOLITAN TRANSIT AUTHORITY   

     BasePay  OvertimePay   OtherPay  Benefits   TotalPay  TotalPayBenefits  \
0  167411.18          0.0  400184.25       0.0  567595.43         567595.43   

   Year         Agency  
0  2011  San Francisco

**¿Cual es el nombre de la persona peor paga (incluyendo beneficios)? ¿Notan algo extraño sobre cuánto le pagan?**

In [0]:
#Copie el Codigo Aca


Id EmployeeName                    JobTitle  BasePay  OvertimePay  \
148653  148654    Joe Lopez  Counselor, Log Cabin Ranch      0.0          0.0   

        OtherPay  Benefits  TotalPay  TotalPayBenefits  Year         Agency  
148653   -618.13       0.0   -618.13           -618.13  2014  San Francisco

**¿Cual es el promedio (mean) BasePay de todos los empleados por año? los años van entre (2011-2014)?**

In [0]:
#Copie el Codigo Aca


Year
2011    63595.956517
2012    65436.406857
2013    69630.030216
2014    66564.421924
Name: BasePay, dtype: float64

**¿Cuantos titulos de trabajos (JobTitle) unicos existen?**

In [0]:
#Copie el Codigo Aca


2159

**Cuales son los 5 trabajos mas comunes?**

In [0]:
#Copie el Codigo Aca


Transit Operator                7036
Special Nurse                   4389
Registered Nurse                3736
Public Svc Aide-Public Works    2518
Police Officer 3                2421
Name: JobTitle, dtype: int64

**¿Cuántos puestos de trabajo fueron representados por una sola persona en 2013? (por ejemplo, los títulos de trabajo con solo una ocurrencia en 2013?)**

In [0]:
#Copie el Codigo Aca


202

**¿Cuántas personas tienen la palabra Chief en el título de su trabajo? (Esta pregunta es mas un poco mas compleja)**

In [0]:
#Copie el Codigo Aca


627

**Bono: ¿Existe una correlación entre la longitud del nombre del Título del trabajo y el Salario?**

Utilice el metodo `.corr()` de Pandas para calcular la correlacion

In [0]:
#Copie el Codigo Aca


title_len  TotalPayBenefits
title_len          1.000000         -0.036878
TotalPayBenefits  -0.036878          1.000000